plugin-gradle/src/main/java/com/diffplug/gradle/spotless/SpotlessPlugin.java

```diff
@@ -107,10 +107,12 @@ public Object doCall(TaskExecutionGraph graph) {
 					.all(task -> task.dependsOn(rootCheckTask));
 		}
 
-		// clear spotless' cache when the user does a clean
+		// clear spotless' cache when the user does a clean, but only after any spotless tasks
+		Task clean = project.getTasks().getByName(BasePlugin.CLEAN_TASK_NAME);
+		clean.doLast(unused -> SpotlessCache.clear());
 		project.getTasks()
-				.matching(task -> task.getName().equals("clean"))
-				.all(task -> task.doLast(unused -> SpotlessCache.clear()));
+				.matching(task -> task instanceof SpotlessTask)
+				.all(task -> clean.mustRunAfter(task));
```

Comment:
@nedtwigg IIUC from what you said at https://github.com/diffplug/spotless/issues/159#issuecomment-339445006, this line ought to be, `.all(task -> task.mustRunAfter(clean))`. Have I misunderstood something here?

Reply:
It's also unclear to me if `clean.mustRunAfter(task)` forces `clean` to run regardless of whether or not a user requested it on top of `spotless{Apply,Check}`.

In other words, do `./gradlew spotlessCheck` and `./gradlew {clean} spotlessCheck {clean}` do the same thing now?

Reply:
`mustRunAfter` is [not the same as `dependsOn`](https://docs.gradle.org/current/javadoc/org/gradle/api/Task.html#setMustRunAfter(java.lang.Iterable))

> this line ought to be, `.all(task -> task.mustRunAfter(clean))`

Good catch!  When I wrote this code, I meant for the order to be `spotless` then `clean`, but that's silly.  "Run the compiler, then clean the output".  lol.  I'll push a fix.

Reply:
> `mustRunAfter` is not the same as `dependsOn`

Cool! Thanks for the clarification.

> Good catch!

You're very welcome!

> When I wrote this code, I meant for the order to be spotless then clean, but that's silly. "Run the compiler, then clean the output". lol. I'll push a fix.

haha, SGTM. :)